Terraform Folder Restructure

starkmatt opened the pull request:
<h1>Terraform Code</h1>
As per Kiko's recommendation I've taken a clone of the Repo and restructured the folder so all the files are in one folder.

<h2>VPC</h2>
- Has been approved and merged to master already.

<h2>RDS Aurora</h2>
- Pending review, looks good from my end but please test for your selves.
After we have this sorted we will be able to connect our WP Instance to the DB.

<h2>ECS</h2>
- Still a WIP, having issues with the code.
Need to figure out why applying is failing.


I also need to split the ECS.tf File apart as there are alot of mixed resources in there, I plan to create;

- IAM File
- ASG File
- Cloud Watch File
- Security Group File

And finally an ECS File.

starkmatt committed:
Cloned repo, folder restructure

starkmatt committed:
white spacing issues

deniseddi commented on terraform/variables.tf:
```diff
@@ -0,0 +1,88 @@
+
+variable "private_subnet-wp-a" {
+    type    = string
+    default = "10.0.63.0/18"
```
Because /24 cidr is required for each subnet, we should instead have:
default = "10.0.63.0/24"

starkmatt replied:
Hey Denise,
Thanks for that appreciated.
When I was doing the subneting calculations of the /16 CIDR block I thought /18 would of been an optimal splits of IP's over 4 subnets.

Just tested /24 on the 10.0.0.0/16 CIDR and it worked fine with no issues on AWS !

deniseddi commented on terraform/variables.tf:
```diff
@@ -0,0 +1,88 @@
+
+variable "private_subnet-wp-b" {
+    type    = string
+    default = "10.0.127.0/18"
```
default = "10.0.127.0/24"

deniseddi commented on terraform/variables.tf:
```diff
@@ -0,0 +1,88 @@
+
+variable "public_subnet-wp-a" {
+    type    =  string
+    default = "10.0.191.0/18"
```
default = "10.0.191.0/24"

deniseddi commented on terraform/variables.tf:
```diff
@@ -0,0 +1,88 @@
+
+variable "public_subnet-wp-b" {
+    type    =  string
+    default = "10.0.255.0/18"
```
default = "10.0.255.0/24"

deniseddi committed:
Update variables.tf 

Added the /24 cidr to each of the subnets, instead of /18.

deniseddi committed:
Added NACL resource to the file.

Added two Network ACLs resources, one associating with the public subnets and one for the private subnets. 
Rules and ports can be defined later on.

deniseddi committed:
Update vpc.tf

deniseddi committed:
Added NACL's

Added 2 Network ACL to the file, one associating with the private subnets and one with the public subnets. 
Tested the code, all is working well. 
Note , on CLI after entering "terraform plan" two messages appear, 
- var.name
  Base name to use for resources in the module
  Enter a value: da-wordpress
- var.vpc_id
VPC ID to create cluster in
Enter a value: aws_vpc.da-wordpress-vpc.id

drii-cavalcanti reviewed (commented):
Denise, could you put all subnet CIDRs together into a variable with list as type?

For example:
```
# public subnets

variable "subnet_cidrs_public" {
  description = "Subnet CIDRs for public subnets"
  default = ["10.0.10.0/24", "10.0.20.0/24", "10.0.30.0/24"]
  type = list(string)
}

# private subnets

variable "subnet_cidrs_private" {
  description = "Subnet CIDRs for private subnets"
  default = ["10.0.11.0/24", "10.0.21.0/24", "10.0.31.0/24"]
  type = list(string)
}
```

drii-cavalcanti commented on terraform/vpc.tf:
```diff
@@ -0,0 +1,117 @@
+resource "aws_subnet" "private-wp-a" {
+  vpc_id                  = "${aws_vpc.da-wordpress-vpc.id}"
+  cidr_block              = "${var.private_subnet-wp-a}"
+  availability_zone       = "${data.aws_availability_zones.iac-azs.names[0]}"
+
+  tags = {
+    Name = "private-subnet-a"
+  }
+}
+
+resource "aws_subnet" "private-wp-b" {
+  vpc_id                  = "${aws_vpc.da-wordpress-vpc.id}"
+  cidr_block              = "${var.private_subnet-wp-b}"
+  availability_zone       = "${data.aws_availability_zones.iac-azs.names[1]}"
+
+  tags = {
+    Name = "private-subnet-b"
+  }
+}
+
+resource "aws_subnet" "public-wp-a" {
+  vpc_id                  = "${aws_vpc.da-wordpress-vpc.id}"
+  cidr_block              = "${var.public_subnet-wp-a}"
+  availability_zone       = "${data.aws_availability_zones.iac-azs.names[2]}"
+
+  tags = {
+    Name = "public-subnet-a"
+  }
+}
+
+resource "aws_subnet" "public-wp-b" {
+  vpc_id                  = "${aws_vpc.da-wordpress-vpc.id}"
+  cidr_block              = "${var.public_subnet-wp-b}"
+  availability_zone       = "${data.aws_availability_zones.iac-azs.names[0]}"
+
+  tags = {
+    Name = "public-subnet-b"
+  }
+}
+
```
Here we can optimize the syntax as well:

```
####### define the subnets #######
resource "aws_subnet" "public-subnet-iac" {
  count = "${length(var.subnet_cidrs_public)}"

  vpc_id = "${aws_vpc.devopsacademy-vpc-iac.id}"
  cidr_block = "${var.subnet_cidrs_public[count.index]}"
  availability_zone = "${data.aws_availability_zones.available.names[count.index]}"

  tags = {
    Name = "public-${data.aws_availability_zones.available.names[count.index]}"
  }
}
resource "aws_subnet" "private-subnet-iac" {
  count = "${length(var.subnet_cidrs_private)}"

  vpc_id = "${aws_vpc.devopsacademy-vpc-iac.id}"
  cidr_block = "${var.subnet_cidrs_private[count.index]}"
  availability_zone = "${data.aws_availability_zones.available.names[count.index]}"

  tags = {
    Name = "private-${data.aws_availability_zones.available.names[count.index]}"
  }
}
```

deniseddi replied:
Hi Adriana, at this stage I am focusing on my task, which is to meet the requirements for the VPC (main file) making sure that works and post the solution asap. I am sure there are plenty of changes that can be made later on.

drii-cavalcanti commented:
Finally, could you create a repository structure?

something like: 

Terraform:
├── README.md
├── variables.tf
├── main.tf
├── outputs.tf
├── modules
│   ├── VPC
│   │   ├── README.md
│   │   ├── variables.tf
│   │   ├── vpc.tf
│   │   ├── outputs.tf
│   ├── ECS
│   │   ├── README.md
│   │   ├── variables.tf
│   │   ├── ecs.tf
│   │   ├── outputs.tf

faria-marcio commented:
I could run the terraform successfully, but the terraform destroy return the following error:
- Error: RDS Cluster FinalSnapshotIdentifier is required when a final snapshot is required

In the exercise c04-rds02 we had to delete a database, so it might be worth have a look at it.

Apart from that, the terraform code LGTM. But I am wondering if this PR isn't duplicated as the VPC terraform code was already merged to master in a different PR (I guess), and there are the tasks #14 RDS - DB and #8 ECS Cluster to delivery these terraform's code?

starkmatt commented:
> I could run the terraform successfully, but the terraform destroy return the following error:
> 
> * Error: RDS Cluster FinalSnapshotIdentifier is required when a final snapshot is required
> 
> In the exercise c04-rds02 we had to delete a database, so it might be worth have a look at it.
> 
> Apart from that, the terraform code LGTM. But I am wondering if this PR isn't duplicated as the VPC terraform code was already merged to master in a different PR (I guess), and there are the tasks #14 RDS - DB and #8 ECS Cluster to delivery these terraform's code?

Hey Marcio,
I left a note here regarding the FinalSnapshotIdentifier

https://github.com/devopsacademyau/2020-feb-project1-group2/pull/15

<em>PS
You may run into some issues when running terraform destroy after.
If this is the case, you'll need to edit your tf.state file and set the following argument to true
skip_final_snapshot = true</em>

Yes the code is duplicated, we decided on putting the code into all one folder so it's easier to add and contribute to the source code collectively 😄

starkmatt committed:
Delete user_data.sh

not needed in this PR

starkmatt committed:
Delete wp.json

not needed in this PR

deniseddi reviewed (approved):
let's start merging it!
